### PR TITLE
Adding BlockSparseTensor support

### PR DESCRIPTION
**Pull request by mganahl**
First parts of BlockSparseTensor support

### CODE REVIEW

**Comment by Thenerdstation** [comment]
I wounder why there's always repeated commits with your PRs. It doesn't matter at all since I squash and merge anyway, just curious...

File: tensornetwork/block_sparse/block_tensor.py [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+  """
+  Flatten a list of lists into a single list.
+  Args:
+    list_of_lists: A list of lists.
+  Returns:
+    list: The flattened input.
+  """
+  res = []
+  for l in list_of_list:
+    res.extend(l)
```
Comment (Thenerdstation) [inline tensornetwork/block_sparse/block_tensor.py]: I did not know this function existed. Cool!

File: tensornetwork/block_sparse/block_tensor.py [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+      np.arange(0, strides[n] * dims[n], strides[n], dtype=np.uint32)
+      for n in range(len(dims))
+  ])
+
+
+def compute_sparse_lookup(charges: List[BaseCharge], flows: List[bool],
+                          target_charges: BaseCharge
+                         ) -> Tuple[np.ndarray, np.ndarray, np.ndarray]:
+  """
+  Compute lookup table for looking up how dense index positions map
```
Comment (Thenerdstation) [inline tensornetwork/block_sparse/block_tensor.py]: nit: change to "Computer lookup table for how dense index positions..."

File: tensornetwork/block_sparse/block_tensor.py [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+    np.ndarray: The degeneracies of each unqiue fused charge.
+  """
+  if len(charges) == 1:
+    return (charges[0] * flows[0]).unique(return_counts=True)
+
+  # get unique charges and their degeneracies on the first leg.
+  # We are fusing from "left" to "right".
+  accumulated_charges, accumulated_degeneracies = (charges[0] *
+                                                   flows[0]).unique(
+                                                       return_counts=True)
```
Comment (Thenerdstation) [inline tensornetwork/block_sparse/block_tensor.py]: nit:
```python
accumulated_charges, accumulated_degeneracies = (
  charges[0] * flows[0]).unique(return_counts=True)
```
is more readable.

File: tensornetwork/block_sparse/block_tensor.py [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+  for n in range(1, len(charges)):
+    leg_charges, leg_degeneracies = charges[n].unique(return_counts=True)
+    # print(accumulated_charges.unique_charges)
+    # print(accumulated_charges.charge_labels)
```
Comment (Thenerdstation) [inline tensornetwork/block_sparse/block_tensor.py]: Remove commented out code

**Comment by mganahl** [comment]
> I wounder why there's always repeated commits with your PRs. It doesn't matter at all since I squash and merge anyway, just curious...

Hmm, could be because I have a bunch of local branches that depend on each other, e.g. i started with add_charge_1 to add some stuff, then from add_charge_1 create a new branch add_charge_2 which adds more stuff a.s.o. That way I can locally merge any requested changes to the PR up the chain to the one PR I'm preparing...

**Comment by Thenerdstation** [comment]
That would make the most sense. In any case it has no effect.

File: tensornetwork/block_sparse/block_tensor.py (hunk 3 above)
Comment (mganahl) [reply tensornetwork/block_sparse/block_tensor.py]: yeah, my yapf configuration in emacs somehow makes these terrible line breaks. Haven't figured out how to change this yet

**Comment by mganahl** [comment]
should be good

**Pull request merged**